u-root.go

```diff
@@ -68,17 +74,92 @@ func init() {
 
 	noCommands = flag.Bool("nocmd", false, "Build no Go commands; initramfs only")
 
-	flag.Var(&extraFiles, "files", "Additional files, directories, and binaries (with their ldd dependencies) to add to archive. Can be speficified multiple times.")
+	flag.Var(&extraFiles, "files", "Additional files, directories, and binaries (with their ldd dependencies) to add to archive. Can be specified multiple times.")
+	statsOutputPath = flag.String("stats-output-path", "", "Write build stats to this file (JSON)")
+	statsLabel = flag.String("stats-label", "", "Use this statsLabel when writing stats")
+}
+
+type buildStats struct {
+	Label      string  `json:"label,omitempty"`
+	Time       int64   `json:"time"`
+	Duration   float64 `json:"duration"`
+	OutputSize int64   `json:"output_size"`
+}
+
+func writeBuildStats(stats buildStats, path string) error {
+	var allStats []buildStats
+	if data, err := ioutil.ReadFile(*statsOutputPath); err == nil {
+		json.Unmarshal(data, &allStats)
+	}
+	found := false
+	for i, s := range allStats {
+		if s.Label == stats.Label {
+			allStats[i] = stats
+			found = true
+			break
+		}
+	}
+	if !found {
+		allStats = append(allStats, stats)
+		sort.Slice(allStats, func(i, j int) bool {
+			return strings.Compare(allStats[i].Label, allStats[j].Label) == -1
+		})
+	}
+	data, err := json.MarshalIndent(allStats, "", "  ")
+	if err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(*statsOutputPath, data, 0644); err != nil {
+		return err
+	}
+	return nil
+}
+
+func generateLabel() string {
+	var baseCmds []string
+	if len(flag.Args()) > 0 {
+		// Use the last component of the name to keep the label short
+		for _, e := range flag.Args() {
+			baseCmds = append(baseCmds, path.Base(e))
+		}
+	} else {
+		baseCmds = []string{"core"}
+	}
+	return fmt.Sprintf("%s-%s-%s-%s", *build, runtime.GOOS, runtime.GOARCH, strings.Join(baseCmds, "_"))
 }
 
 func main() {
 	flag.Parse()
 
+	start := time.Now()
+
 	// Main is in a separate functions so defers run on return.
 	if err := Main(); err != nil {
 		log.Fatal(err)
 	}
-	log.Printf("Successfully wrote initramfs.")
+
+	elapsed := time.Now().Sub(start)
+
+	stats := buildStats{
+		Label:    *statsLabel,
+		Time:     start.Unix(),
+		Duration: float64(elapsed.Milliseconds()) / 1000,
+	}
+	if stats.Label == "" {
+		stats.Label = generateLabel()
+	}
+	if stat, err := os.Stat(*outputPath); err == nil {
```

Comment:
if what was at outputPath already existed before u-root, and u-root fails, then we'd still print a size... can you fix it so that it only prints a size if we truly wrote something there?

Reply:
i think it's not the case because Main would have failed in that case, but added an mtime check, it's easy enough.

i also realized that we weren't capturing the size of `cmds/exp/*`, as they are not included in any of the templates.
assuming that not including them in `all` is intentional and should not be changed, i created another template. i called it `world` but i'm open to suggestions. `everything`?

Reply:
oh you're right. sorry.